### ISSUE TITLE
require('leaflet') instead of global script tag

### ISSUE THREAD

**Issue by batpad** [Project-OSRM/osrm-frontend-v2#137]
Is there a good reason to include Leaflet globally via a `<script>` tag instead of just requiring it inside index.js with `var L = require('leaflet');` or so?

It seems really messy to mix a global leaflet object and having different modules `require('leaflet')` as well.

Not sure if there's a reason for this that I'm missing. Am happy to make a pull request to fix this if no one has any objections.

**Comment by amyleew** [comment]
Unless @TheMarex has any objections, I say go for it with a PR @batpad !

**Comment by TheMarex** [comment]
Nope wasn't sure this works, but if it does, go for it.

**Comment by batpad** [comment]
So, this seems like it's a little bit of voodoo, am just going to describe the behaviour I noticed and what my PR contains, and will let you guys take a call on whether it makes sense to merge or not:
1. I just remove the global `<script>` tag without doing anything else and rebuilt - strangely, everything seems to work. This seems wrong though - my guess is it still works because Leaflet gets defined globally anyways with packages requiring it in, but of course, one would want to have an explicit `require` everywhere it is being used.
2. I try adding latest leaflet (0.7.5) to package.json, installing it, and adding `var L = require('leaflet');` in all the files that use Leaflet. This fails with `L.Routing.Plan is undefined`. I notice that leaflet-routing-machine is using leaflet 0.7.2, and _probably_ hence lands up some-how creating a "different" instance of `L` than is available in index.js (or something like that?)
3. I change the version of leaflet in the osrm-frontend-v2 package.json to 0.7.2 to match leaflet-routing-machine and everything _seems_ to work.

So I am making a PR with 3. I have tested on my end and everything _seems_ to work okay, but of course, please test thoroughly :). Let me know if there's anything I can do to improve the PR. Thanks!

**Comment by TheMarex** [comment]
@batpad I would guess that LRM needs to be loaded after leaflet. So you would need to include it by `require` as well for this to work?

/cc @perliedman

**Comment by perliedman** [comment]
Like @batpad said, it might be that two different versions of Leaflet get loaded, one by the frontend and one by LRM. I don't remember the details, but [`npm dedupe`](https://docs.npmjs.com/cli/dedupe) can be used to resolve such issues.

Since LRM tacks on the `Routing` object onto `L`, it's important that just one instance of Leaflet is present through the whole app.

**Comment by batpad** [comment]
@TheMarex lrm is already being loaded via require, so that should be fine? The problem was that LRM internally was requiring an older version of Leaflet - v0.7.2. If I set the version of Leaflet in osrm-frontend-v2 to be v0.7.2, everything seems to work fine and as expected (The L object is the same between osrm-frontend-v2 and lrm and L.Routing gets added correctly).

I have tested with my PR above and everything seems to work fine (you will need to `npm install` after pulling the code). The disadvantage here is we are down-grading to Leaflet v0.7.2, not sure if that is a cause for concern. Of course, ideally LRM would just update its version of Leaflet and then we could do the same.

Let me know if anyone has a chance to test out the PR and if something doesn't work? @perliedman @TheMarex do you see any reason this might not be a good idea?

Thanks!

**Comment by perliedman** [comment]
Hm, dependency management in npm isn't my core expertise, but LRM sets `^0.7.2`, so it should accept 0.7.5 as well, and I think npm will download 0.7.5 if you start with a clean build?

LRM could of course specify 0.7.5, but since it does in fact work with 0.7.2 (and probably all 0.7 versions), this would at the same time prevent using LRM with earlier versions, which you might want for various reasons.

**Comment by perliedman** [comment]
I verified this - npm will install latest acceptable version if you install LRM from scratch:

```
$ npm install --save leaflet-routing-machine
leaflet-routing-machine@2.5.0 node_modules\leaflet-routing-machine
├── corslite@0.0.6
├── polyline@0.0.3
└── leaflet@0.7.5
```

**Comment by batpad** [comment]
@perliedman right. I did an `rm -rf node_modules` and then specifying leaflet ^0.7.5 as dependency seems to fetch the right version all around and everything works. Have updated the pull request to just use leaflet 0.7.5. So, everything seems to work for me this way, without needing the `<script>` tag in index.html. Can someone else test, please? If it works, do you see any other potential problems, or would this be possible to merge?

Thanks!

**Comment by freenerd** [comment]
Fixed with #184
